Pass well-defined set of managed assemblies to linker and crossgen

nguerrera opened the pull request:
### Description

Both the linker and crossgen features currently scan the entire set of published files, looking for managed assemblies. This was meant to be temporary and the design was to pass them the set of managed assemblies designated in .deps.json. This was not trivial due to ordering issues so we have been living with the temporary design. This change implements the planned design of passing the true set of managed assemblies.

The temporary design is inefficient, there's a first chance exception for every non-dll in the publish set, and can be incorrect in the edge case where there are dlls that are not part of the applications runtime dependencies included as content.

While working on this, bugs related to identifying the correct set of assemblies in the .deps file were fixed. It is difficult to split up the bug fixes as they were done on the foundation of the better design.

In total, 4 bugs are fixed.

1. Fix #3109 -  tracked the correct design that is implemented here
2. Fix #3547 - COMReference with PIA do not show up in .deps file, and therefore do not load. EmbedInteropTypes must be used.
3. Fix #2660 - Project references that are not copy local are incorrectly recorded in .deps file
4. Fix #3548 - linker cannot remove project references or direct file references from deps file

Relates to a fifth bug, https://github.com/aspnet/AspNetCore/issues/12064 (failure to account for views assembly in linker and crossgen), but a fix for that has been made that does not depend on this one. It includes a future proofing so that it can work with both the existing design and the design implemented here.

### Customer impact

This is primarily about having the planned correct design in place for the long term so that we do not have to support the temporary design indefinitely. It would be rare for an application to publish incorrectly with the temporary design.

The second bug has a simple workaround of using No-PIA, but the failure is not obvious.

The third bug has been worked around by the runtime. It now tolerates missing assemblies in .deps by default. That was also intended to be temporary so the longer we leave the SDK producing the incorrect .deps file, the longer the runtime will have to support this by default.

The fourth bug is benign for now as the linker only removes framework assemblies in this release, but this is expected to change in future releases, and it demonstrates the problem of having the poorer design in place as the foundation for these features evolving.

### Regression

No


### Risk

Medium. The code delta is not too large, but the change is in a tricky part of the code and it took 3 tries to come up with the right approach. More regression testing is still needed, which may take until tomorrow to finish authoring.

nguerrera committed:
Pass well-defined set of managed assemblies to linker and crossgen

Both were taking all the files to publish and checking what was managed code. This
was always intended to be temporary pending this change.

Several bugs fixed in the process, will update this will all the details.

nguerrera commented:
@sbomer @fadimounir Still draft as I'm missing tests, but FYI

nguerrera committed:
Add test for web app going through linker and crossgen

nguerrera committed:
Skip test while razor bug is active

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets:
```diff
@@ -762,7 +761,67 @@ Copyright (c) .NET Foundation. All rights reserved.
+    
+    <!-- Assemblies from runtime packs for self-contained apps -->
+    <CallTarget Targets="_ComputeManagedRuntimePackAssemblies" Condition="'$(SelfContained)' == 'true'">
+      <Output TaskParameter="TargetOutputs" ItemName="_ManagedRuntimeAssembly" />
```
Did you get an answer on whether or not this was the recommended way to conditionally depend on a target?

I think you can also conditionally set a property that you use for DependsOnTargets.

nguerrera replied:
I’ll do that

dsplaisted reviewed (approved):
Is it possible to add tests for the bugs that were fixed?

nguerrera commented:
I am working on the tests and expect to have them up tomorrow

nguerrera committed:
Add test coverage for linker removing project refs

nguerrera committed:
Add COMReference coverage

nguerrera committed:
Add test coverage for non-copy-local project reference

livarcocc commented:
Please, re-target to 3.1.1xx and merge when you are ready.

nguerrera committed:
Fix up comments

nguerrera committed:
Use HashSet to avoid O(N*M)

nguerrera committed:
Replace CallTarget with variable DependsOnTargets

nguerrera commented:
@sbomer, @fadimounir Please review. This will go in 3.1 and not 3.0.

nguerrera commented:
@dsplaisted I made the requested changes and added test coverage for all the fixed bugs if you want to take another look.

nguerrera commented:
Sorry the retarder to 3.1 pulled in unrelated commits. I will fix this in the morning

nguerrera commented:
Changed base back to 3.0 for now so this can be reviewed. Will submit 3.0 -> 3.1 merge separately then retarget again.

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/RunReadyToRunCompiler.cs:
```diff
@@ -19,7 +20,7 @@ public class RunReadyToRunCompiler : ToolTask
         public ITaskItem CompilationEntry { get; set; }
         [Required]
-        public ITaskItem[] ImplementationAssemblies { get; set; }
+        public ITaskItem[] References { get; set; }
```
I think the term `References` here could add ambiguities in the future, and someone could end up passing non-implementation assemblies as references to crossgen. I would like the names to reflect the notion of "implementation assemblies" used as crossgen references here if possible. Maybe something like `ImplementationAssemblyReferences`?

nguerrera replied:
I'll rename this. Just implementation assemblies was confusing me as to whether it was for referencing or compiling.

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -112,20 +115,33 @@ protected override void ExecuteCore()
+                    continue;
+                }
+
+                r2rReferenceList.Add(file);
```
I don't think this is correct because the `GetInputFileEligitility` now returns `Eligibility.ReferenceOnly` for reference assemblies, and we can't reference those. Crossgen needs to reference implementation assemblies.

fadimounir replied:
I'm thinking of the case where we get those references assemblies published under the refs folder. Just wanting to make sure they don't show up here

nguerrera replied:
The refs folder would not be here because this is no longer just everything in in resolved files to publish. The set of assemblies are supposed to be known implementation assemblies before we get this far. But I decided to be careful and still check that they are actually assemblies. So it makes sense to be careful there too.

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -221,32 +232,47 @@ bool InputFileEligibleForCompilation(ITaskItem file)
                         }
 
                         // Skip reference assemblies and assemblies that reference winmds
                         if (ReferencesWinMD(mdReader) || IsReferenceAssembly(mdReader) || !HasILCode(pereader, mdReader))
```
We should probably return `Eligibility.None` when `IsReferenceAssembly` returns true to prevent crossgen from referencing non-implementation assemblies.

nguerrera replied:
OK, this should not happen, but it cannot hurt to be defensive and checking Eligibility at all for references is defensive. The list coming in should be all managed implementation assemblies already.

fadimounir commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -221,32 +232,47 @@ bool InputFileEligibleForCompilation(ITaskItem file)
             }
 
-            return true;
+            if (file.HasMetadataValue(MetadataKeys.ReferenceOnly, "true"))
```
Just to confirm here, again, we need to make sure that the `ReferenceOnly` metadata does not get added to reference assemblies, since crossgen can't take those as references

fadimounir reviewed (changes requested):
The code generally looks ok to me (I only reviewed the crossgen related changes), but I'm suspecting there could be a bug that would make it possible to give reference assemblies to crossgen, and crossgen needs to reference implemenation assemblies, not reference assemblies

nguerrera committed:
Skip COMReference test because it depends on machine state

nguerrera committed:
Defend against reference assemblies being passed to crossgen as references

Also re-order checks for perf

nguerrera committed:
Mark methods private explicitly, and static where possible

nguerrera committed:
Remove unused method

nguerrera committed:
Rename References to ImplementationAssemblyReferences

This to re-enforce for the code reader that crossgen references,
must be implementation assemblies.

nguerrera commented:
@fadimounir Feedback addressed. Please take another look.

fadimounir reviewed (approved):
LGTM. Thanks for fixing these issues

johnbeisner commented on src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs:
```diff
@@ -49,10 +49,13 @@ public class PrepareForReadyToRunCompilation : TaskBase
         public ITaskItem[] ReadyToRunFilesToPublish => _r2rFiles.ToArray();
 
+        [Output]
+        public ITaskItem[] ReadyToRunFilesToReference => _r2rReferences.ToArray();
```
```suggestion
        public ITaskItem[] ReadyToRunAssembliesToReference => _r2rReferences.ToArray();
```

johnbeisner replied:
This should fix the failing tests. Running tests on my dev box now...

nguerrera replied:
Thanks, that was a last minute change and I haven't gotten back to this PR since I returned from vacation yesterday.

nguerrera committed:
Update src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs

Co-Authored-By: John Beisner <johnbeisner@users.noreply.github.com>
